perf: refactor SIP-68 db migrations with INSERT SELECT FROM

ktmud opened the pull request:
### SUMMARY

This is another take on #19406 and #19416. The goal is to rewrite the bulk of loading + rewriting operations from Python to native SQL statements by utilizing `INSERT SELECT FROM`.

The whole migration happens in 3 steps:

1. Copy `tables`, `table_columns`, `sql_metrics` to the new `sl_xxx` tables using INSERT SELECT FROM.
2. Copy model associations to relationship tables.
3. Run post processing on `sl_datastes` and `sl_columns` to
   - Tuck additional metadata columns (`verbose_name`, `d3format`, etc) in the original tables under `extra_json`
   - Add conditional quotes to physical table and column expressions
   - Apply dataset level `is_managed_externally` and `external_url` to columns

Comparing to #19406, this PR was able to cut down the migration time for our Superset instance from 7 hours to 1 hour while retaining the full functionality.

While working on the migration, I also noticed a couple of shortcomings in the original implementation, including problems in the shadow-writing logic. Most notably:

1. Some long text columns should use `MediumText` for MySQL
2. Columns are not automatically deleted when they are unlinked from a table or dataset (e.g. when you modify or remove a column from a dataset)
3. Letting physical table and dataset point to the same columns could be problematic as we wouldn't want users who updated column properties in a dataset to automatically propagate that change to the related table. Tables and datasets may have different ownerships and different users may create different physical datasets for different tables. One of the original goal of SIP-68 was to change the mapping between physical datasets and database table/views to an n:1 relationship. It only makes sense if column definitions in each dataset are independent as well.
4. Dataset does not know which database it belongs to unless being joined with a table. However, a dataset does not always have an associated table---for example, we can have virtual queries like "SELECT 1;" that does not populate data from any tables. Or, the table name extraction for a complex SQL statement could fail (especially when it has Jinja syntax).

Instead of creating multiple db migrations to address these issues, it'd be much easier for end users if we just skip the original migration (drop data for who already migrated) and bundle everything in another migration:

1. Datasets and columns now reuse uuids from the old tables. This makes filling up the association tables easier. It also makes it possible to import a Superset export from an old Superset instance that hasn't upgrade to the new models, even after the old models have been removed.
2. `created_on`, `change_on`, `created_by` and `changed_by` are also copied over when appropriate.
3. In MySQL, some text columns (`expression`, `description`, etc) are now `MediumText`. This is consistent with current column types in the old tables.
4. Since old and new entities can now be linked via `uuid`, `sqlatable_id` on `NewDataset` is removed. However, there is no 1:1 mapping of `NewTable` in the old models so they acquires uuids. In order to link `NewTable` with associated columns, `sqlatable_id` is added to `NewTable` during the db migration, but dropped once migration is complete.
5. Refactored shadow-writing hooks to reduce duplicate code and more appropriately handle column deletion.
6. Removed column fetching for tables. We used to fetch table metadata (column types, etc) from datasources during db migration, as well as in shadowing writing, but this has a major risk---external datasources can be unreliably slow. Users could fail the migration or saving edits to a datasource if the datasource is temporarily down. We should not let this unreliable process block the critical path of basic functionalities. I added a `Table.sync_columns` method, but in the future this should be consolidated with `get_virtual_table_metadata` and `get_physical_table_metadata`.
7. `dataset.owners` are also added (copying PR #19487).

#### Known issues and next steps

1. The table columns for `NewTable` will only be copied from the dataset when its being created. We'd need another mechanism to sync table columns. Either during re-populating after `fetch_metadata` or in an async script. Since the Table model is not used anywhere yet, I think it's okay to leave this in another followup PR.
2. I added a [`sync_columns`](https://github.com/apache/superset/blob/03295342428faff6958f7a4a1181ac69bb88c6fd/superset/tables/models.py#L122) method for tables. It's not currently in use and does more or less the same as `SqlaTable.fetch_metadata`. I think it's okay to keep this somewhat-duplicate not-in-active-use method as `SqlaTable` will be remove eventually.
3. In this PR, `NewTable.columns` will be updated only when a **physical** `SqlaTable` is first synced to the new models and a `NewTable` is created. Updating columns in a `SqlaTable` dataset will NOT update `SqlaTable.columns` as we consider table columns and dataset columns to be separate and the metadata syncing should only happen in one direction---that is from tables to datasets, not the other way around.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

<img width="600" alt="image" src="https://user-images.githubusercontent.com/335541/161164424-b12d34ce-b7ab-4a6b-a6a1-b0ca2b25f649.png">


### TESTING INSTRUCTIONS

To test the migration

```shell
superset db downgrade 2ed890b36b94 && SHOW_PROGRESS=1 superset db upgrade a9422eeaae74
```

A couple of queries to verify the correctness of the migrated data:

```sql
select
  sl_tables.name as table_name,
  sl_datasets.name as dataset_name,
  sl_datasets.expression
from sl_datasets
  join sl_dataset_tables on sl_datasets.id = sl_dataset_tables.dataset_id
  join sl_tables on sl_tables.id = sl_dataset_tables.table_id
limit 5;
```

```sql
select id, name, extra_json, changed_by_fk
from sl_columns
where
  changed_by_fk is not null
  and extra_json like "%verbose_name%"
limit 10;
```

```sql
select
  c.id, c.name, c.expression, c.uuid, c.created_on
from sl_columns as c join sl_dataset_columns a
  on a.column_id = c.id
where a.dataset_id = <your id>
order by created_on desc limit 10;
```

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: SIP-68, #14909 and #17543 
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

ktmud committed:
refactor: rewrite new dataset models with INSERT FROM

john-bodley commented on superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:
```diff
@@ -292,78 +305,70 @@ def after_insert(target: SqlaTable) -> None:  # pylint: disable=too-many-locals
-                expression=metric.expression,
-                warning_text=metric.warning_text,
                 description=metric.description,
+                expression=metric.expression,
```
Oh my. I do love me some ABC.

ktmud replied:
I know this would make you happy, John! But unfortunately I think this block will be eventually removed once we migrate to INSERT FROM SELECT.

john-bodley commented on superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:
```diff
@@ -292,78 +305,70 @@ def after_insert(target: SqlaTable) -> None:  # pylint: disable=too-many-locals
         predicate = or_(
             *[
                 and_(
+                    NewTable.database_id == database_id,
```
Nice!

serenajiang commented on superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:
```diff
@@ -207,427 +244,557 @@ class NewTable(Base):
+            if verbose_name is not None:
+                updated_extra["verbose_name"] = verbose_name
+            if python_date_format is not None:
+                updated_extra["python_date_format"] = verbose_name
```
```suggestion
                updated_extra["python_date_format"] = python_date_format
```
copy paste error?

ktmud replied:
Thanks for the catch! I changed this to a dict to avoid such typo.

eschutho commented on superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:
```diff
@@ -150,55 +176,66 @@ def fetch_columns_and_metrics(self, session: Session) -> None:
     Base.metadata,
     sa.Column("table_id", sa.ForeignKey("sl_tables.id")),
     sa.Column("column_id", sa.ForeignKey("sl_columns.id")),
+    UniqueConstraint("table_id", "column_id"),
```
For people who have already run this migration, we won't get anything new that's added in this update. Can we instead create a new migration for any additions or changes that won't affect performance?

eschutho reviewed (dismissed):
Per comment above, would recommend pulling out anything that has changed  in the db structure that doesn't affect performance into a new migration.

ktmud commented:
@eschutho I'm proposing to change current migration to no-op and move my updated code to a new migration. 

See an earlier message I sent in Slack:

---

Hi, I noticed we are making more adjustments to SIP-68 models and have prepared a [couple](https://github.com/apache/superset/pull/19425) of [more](https://github.com/apache/superset/pull/19487) db migrations. I’m wondering whether we should bundle all these migrations (including the first one that’s already merged) into one new migration and change the original migration to no-op.

**Pros:**

- Reduced total migration time: bundle everything should be faster than running them separately
- We get a chance to fix a couple of more errors such as [using MediumText for MySQL](https://github.com/apache/superset/pull/19421#discussion_r839942807) and [incorrect additive_metric_types matching](https://github.com/apache/superset/pull/19421#discussion_r839903477)
- We get a chance to copy over other missing data such as [changed on and last updated](https://github.com/apache/superset/pull/19421#discussion_r840089807)
- We can re-ID the copied entities to follow the original ones, making it easier to spot-check potential data inconsistency bugs down the road
- Everyone’s db is in a clean and consistent state

**Cons:**
- Those who already ran the migration and bore the slowness may have to experience it again

Happy to incorporate [#19487](https://github.com/apache/superset/pull/19487/) and [#19425](https://github.com/apache/superset/pull/19425) to [my PR](https://github.com/apache/superset/pull/19421) if they are still needed. (edited)

eschutho commented:
@ktmud I believe even with changing this PR to have the initial migration be a no-op, the create table functions will fail if the table already exists, unless you pass something like `checkfirst=True`. But then we still won't be getting any new columns or indexes.

eschutho commented:
>Btw, I think the Dataset model may need a database_id column as well. There is the implicit assumption that a dataset can only run on one database. I cannot imagine a case where we need to support a virtual dataset being used on different tables in different databases. Having direct link to databases makes sure existing virtual datasets can be linked to the correct database without relying on an unreliable table name extraction process. Currently if table name extraction fails, a virtual dataset lost its association with a correct table, hence the only link to database. It would require joining SqlaTable with sqlatable_id to get the correct database id.

I believe the having the db id on the table will be important for future features where we need to power a chart by a Table without a dataset, and I'm wary of having the db id in both places in case they become out of sync. I'm not sure if I follow the use case of a table name extraction failing and then the dataset doesn't have a relationship to a db. Doesn't a virtual dataset break if the table extraction doesn't work anyway? What would be the value of having a link to the db but not the table?

ktmud commented:
> @ktmud I believe even with changing this PR to have the initial migration be a no-op, the create table functions will fail if the table already exists, unless you pass something like checkfirst=True. But then we still won't be getting any new columns or indexes.

The new migrations [will drop existing tables ](https://github.com/apache/superset/blob/c21e3d19456b055e672634f1938a009d71dcf747/superset/migrations/versions/a9422eeaae74_new_dataset_models_take_2.py#L802) first. This line does not throw an error if the tables do not exist.


>  I'm wary of having the db id in both places in case they become out of sync

This is a valid point. I guess one solution would be to always create a `NewTable` without a table name for every database schema and link it to every virtual dataset migrated---without doing the unreliable table name extraction. This of course will not require adding a `NewDataset.database_id` field and may even give us a chance to not do table name extraction at all in this migration---so the speed can improve even further.

> Doesn't a virtual dataset break if the table extraction doesn't work anyway?

That's exactly the point. If table name extraction fails, a virtual dataset will become unusable. However, currently the use of a virtual dataset based on `SqlaTable` does not rely on table name extraction. A `SqlaTable` has a direct link to database so the chart will know which database to run the SQL against.

> What would be the value of having a link to the db but not the table?

Ideally we would want a link just to the tables, but that cannot be reliably automated, can it?

ktmud commented:
Another problem I noticed: when editing a datasource and updating an existing column, instead of updating the original one, a new `sl_column` record was created.

ktmud commented:
There is a lot more work involved than original anticipated. Converting to draft for now.

ktmud commented on superset/columns/models.py:
```diff
@@ -52,6 +52,38 @@ class Column(
+    is_spatial = sa.Column(sa.Boolean, default=False)
+
+    # Is this a time column? Useful for plotting time series.
+    is_temporal = sa.Column(sa.Boolean, default=False)
```
A-Z reordering

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -1901,9 +1903,7 @@ def update_table(  # pylint: disable=unused-argument
-            session.query(NewDataset)
-            .filter_by(sqlatable_id=target.table.id)
-            .one_or_none()
+            session.query(NewDataset).filter_by(uuid=target.table.uuid).one_or_none()
```
We now link old and new datasets by uuid instead of a `sqlatable_id` field. Note since `NewDataset` is designed to be a 1:1 replacement for SqlaTable, it should be okay for them to share the same uuid. However, `NewTable` is a totally new entity type, so they should have separate uuids. Therefore a `sqlatable_id` is added to `NewTable` instead.

ktmud reviewed (commented):
More changes to come. Will fix tests later.

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -2063,172 +2169,74 @@ def after_update(  # pylint: disable=too-many-branches, too-many-locals, too-man
+            # add pending new columns to known columns list, too, so if calling
+            # `after_update` twice before changes are persisted will not create
+            # two duplicate columns with the same uuids.
+            dataset.columns = sqla_table.get_sl_columns()
```
Now we just always re-create or update all NewColumn's as long as there is any update to sqla_table.columns or sqal_table.metrics.

eschutho reviewed (approved):
I checked for fixes for all known issues that we've been seeing and they're all covered here.

betodealmeida commented on superset/migrations/versions/a9422eeaae74_new_dataset_models_take_2.py:
```diff
@@ -0,0 +1,858 @@
+def upgrade() -> None:
+    bind = op.get_bind()
+    session: Session = db.Session(bind=bind)
+    Base.metadata.drop_all(bind=bind, tables=new_tables)
+    Base.metadata.create_all(bind=bind, tables=new_tables)
```
Nice, I was going to ask about this.

betodealmeida commented on superset/sql_parse.py:
```diff
@@ -698,3 +707,75 @@ def insert_rls(
+    "clickhouse": {"clickhouse"},
+}
+
+RE_JINJA_VAR = re.compile(r"\{\{[^\{\}]+\}\}")
+RE_JINJA_BLOCK = re.compile(r"\{[%#][^\{\}%#]+[%#]\}")
```
Oh, good catch, I forgot about those.

betodealmeida commented on superset/tables/models.py:
```diff
@@ -80,13 +107,96 @@ class Table(Model, AuditMixinNullable, ExtraJSONMixin, ImportExportMixin):
+            column.is_partition = False  # TODO: update with accurate is_partition
+            return column
+
+        self.columns = list(map(update_or_create_column, column_metadata))
```
Smallest nit:
```suggestion
        self.columns = [update_or_create_column(column) for column in  column_metadata)]
```

betodealmeida commented on superset/tables/models.py:
```diff
@@ -80,13 +107,96 @@ class Table(Model, AuditMixinNullable, ExtraJSONMixin, ImportExportMixin):
+            return []
+
+        if not database.id:
+            raise Exception("Database must be already committed")
```
You get an `id` by calling `session.flush()` as well, I think that's still fine, right?

```suggestion
            raise Exception("Database must be already flushed or committed")
```

betodealmeida reviewed (changes requested):
This is a lot of work, @ktmud! Thanks for bringing this to next level and making sure Superset works at scale!

I left a few comments, there are a few places where I think you need to wrap the call to `NewTable.load_or_create` in a list.

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -2273,95 +2286,54 @@ def write_shadow_dataset(  # pylint: disable=too-many-locals
+                for column in columns
+                if column.is_physical
+            ]
+            tables = NewTable.load_or_create(
```
The function actually returns a list of tables. Maybe I'll rename it to `bulk_load_or_create` to reduce confusion.

betodealmeida replied:
Ah, you're right, and I doubled check while reviewing! I think I got confused by the name like you said...

betodealmeida commented on superset/connectors/sqla/models.py:
```diff
@@ -2237,29 +2245,34 @@ def after_update(  # pylint: disable=too-many-branches, too-many-locals, too-man
+                referenced_tables = extract_table_references(
+                    sqla_table.sql, sqla_table.database.get_dialect().name
+                )
+                dataset.tables = NewTable.load_or_create(
```
Ignore

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -479,6 +541,58 @@ def data(self) -> Dict[str, Any]:
+        target: "SqlMetric",
+    ) -> None:
+        session = inspect(target).session
+        dataset = session.query(NewColumn).filter_by(uuid=target.uuid).one_or_none()
+        if dataset:
+            session.delete(dataset)
```
Thanks for catching this!

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -479,6 +541,58 @@ def data(self) -> Dict[str, Any]:
+        column.changed_on = self.changed_on
+        column.created_by = self.created_by
+        column.changed_by = self.changed_by
+        column.type = "Unknown"
```
I created a constant.

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -1745,13 +1859,16 @@ def fetch_metadata(self, commit: bool = True) -> MetadataResult:
+                    column_name=col["name"],
+                    type=col["type"],
+                    table_id=self.id,
+                    table=self,
```
Yeah, previously I replaced `table` with `table_id` to fix the problem mentioned above but added it back after replacing it with another solution: assigning the columns list to `self.columns` only after all items are collected. I guess I can remove `table_id` again.

ktmud commented on superset/connectors/sqla/models.py:
```diff
@@ -2273,95 +2286,54 @@ def write_shadow_dataset(  # pylint: disable=too-many-locals
-                is_managed_externally=dataset.is_managed_externally,
-                external_url=dataset.external_url,
+        if not self.sql:
+            # always create separate column entities for Dataset and Table
+            # so updating a dataset would not update the columns in the table
```
Just realized I didn't do the same in the migration. Let me fix it.

ktmud committed:
Address PR comments and create duplicate physical columns for tables

ktmud commented:
@betodealmeida I think I've addressed all you comments (and also changed how physical columns for tables were added during migrations). Can you take another look?

betodealmeida reviewed (approved):
This looks great, Jesse!

cemremengu commented:
I get the following error when applying this migration. Even if I update the column to be `False` by default it still does not work. Any ideas?

```
INFO  [alembic.runtime.migration] Running upgrade ad07e4fdbaba -> a9422eeaae74, new_dataset_models_take_2
>> Copy 12 physical tables to sl_tables...
>> Copy 41 SqlaTable to sl_datasets...
   Copy dataset owners...
   Link physical datasets with tables...
>> Copy 2,326 table columns to sl_columns...
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.NotNullViolation: null value in column "is_temporal" of relation "sl_columns" violates not-null constraint
DETAIL:  Failing row contains (368c1f15-f08d-440b-b21d-b113e1a4e42d, 2021-12-03 08:09:13.776918, 2021-12-03 21:20:46.402427, 27, null, f, f, t, t, t, f, f, t, null, f, default_serial_num, STRING, null, , null, null, null, {"warning_markdown":null}, 1, 1).
```

ktmud commented:
@cemremengu thanks for reporting this. I'll work on a fix.

cemremengu commented:
There is another one @ktmud not sure about the fix yet:

```
>> Run postprocessing on 41 datasets
   Process dataset 1~41...


  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.8/site-packages/alembic/runtime/environment.py", line 813, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.8/site-packages/alembic/runtime/migration.py", line 561, in run_migrations
    step.migration_fn(**kw)
  File "/app/superset/migrations/versions/a9422eeaae74_new_dataset_models_take_2.py", line 887, in upgrade
    postprocess_datasets(session)
  File "/app/superset/migrations/versions/a9422eeaae74_new_dataset_models_take_2.py", line 581, in postprocess_datasets
    quoted_expression = get_identifier_quoter(drivername)(expression)
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/sql/compiler.py", line 3700, in quote
    if self._requires_quotes(ident):
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/sql/compiler.py", line 3613, in _requires_quotes
    or value[0] in self.illegal_initial_characters
IndexError: string index out of range

```

ktmud commented:
@cemremengu does changing L580 to 

```python
            if is_physical and drivername and expression:
```

fix the error? It should not happen but it seems somehow your physical columns have empty strings as the `expression`.

cemremengu commented:
I think it is because my connection string is in the form `postgresql+psycopg2://` so sqlalchemy is failing to find an identifier quoter ?

ktmud commented:
No, if it's a connection string error, you won't even reach line 3613 of `sqlalchemy/sql/compiler.py`. Are these all the records in your `sl_datasets` table?

cemremengu commented:
Yes, I removed the + part and still got the error unfortunately.

> Are these all the records in your sl_datasets table?

Yes but it is strange since I have 41 datasets currently. Is there any chance that I did one of the previous migrations incorrectly?

ktmud commented:
@cemremengu that's very strange. can you downgrade and rerun the whole migration?

```
superset db downgrade ad07e4fdbaba && SHOW_PROGRESS=1 superset db upgrade a9422eeaae74
```

cemremengu commented:
Downgrading *combined with*  `if is_physical and drivername and expression:` fixed the issue. Thanks so much!

PS. I had to manually update nextval for the sl_* table sequences but that might be due to my previous attempts.
